redbot/core/tree.py

```diff
@@ -0,0 +1,324 @@
+        If commands are manually added to the tree outside of the standard cog loading process, this must be run
+        for them to be usable.
+        """
+        enabled_commands = await self.client.list_enabled_app_commands()
+
+        to_add_commands = []
+        to_add_context = []
+        to_remove_commands = []
+        to_remove_context = []
+
+        # Add commands
+        for command in enabled_commands["slash"]:
+            if command in self._disabled_global_commands:
+                to_add_commands.append(command)
+
+        # Add context
+        for command in enabled_commands["message"]:
+            key = (command, None, AppCommandType.message.value)
+            if key in self._disabled_context_menus:
+                to_add_context.append(key)
+        for command in enabled_commands["user"]:
+            key = (command, None, AppCommandType.user.value)
+            if key in self._disabled_context_menus:
+                to_add_context.append(key)
+
+        # Remove commands
+        for command in self._global_commands:
+            if command not in enabled_commands["slash"]:
+                to_remove_commands.append((command, AppCommandType.chat_input))
+
+        # Remove context
+        for command, guild_id, command_type in self._context_menus:
+            if guild_id is not None:
+                continue
+            if (
+                AppCommandType(command_type) is AppCommandType.message
+                and command not in enabled_commands["message"]
+            ):
+                to_remove_context.append((command, AppCommandType.message))
+            elif (
+                AppCommandType(command_type) is AppCommandType.user
+                and command not in enabled_commands["user"]
+            ):
+                to_remove_context.append((command, AppCommandType.user))
+
+        # Actually add/remove
+        for command in to_add_commands:
+            super().add_command(self._disabled_global_commands[command])
+            del self._disabled_global_commands[command]
+        for key in to_add_context:
+            super().add_command(self._disabled_context_menus[key])
+            del self._disabled_context_menus[key]
+        for command, type in to_remove_commands:
+            com = super().remove_command(command, type=type)
+            self._disabled_global_commands[command] = com
+        for command, type in to_remove_context:
+            com = super().remove_command(command, type=type)
+            self._disabled_context_menus[(command, None, type.value)] = com
+
+    @staticmethod
+    async def _send_from_interaction(interaction, *args, **kwargs):
+        """Util for safely sending a message from an interaction."""
+        ctx = await interaction.client.get_context(interaction)
+        try:
+            await ctx.send(*args, ephemeral=True, **kwargs)
+        except discord.HTTPException:
+            await ctx.channel.send(*args, **kwargs)

```

Comment:
Note this can be done without context:

```suggestion
        if interaction.response.is_done():
            if interaction.is_expired():
                return await interaction.channel.send(*args, **kwargs)
            return await interaction.followup.send(*args, ephemeral=True, **kwargs)
        return await interaction.response.send_message(*args, ephemeral=True, **kwargs)
```